### PR TITLE
Add dynamic filterable repository. (fixes #11)

### LINKED ISSUE #11: Add paging ability to repository
ITNOA

Paging feature is good ability for repositories. So it is good to have it ;)

### PR DESCRIPTION
**Pull request by soroshsabz**
ITNOA

This PR close #11

### CODE REVIEW

File: Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    using BSN.Commons.Orm.EntityFrameworkCore.Extensions;
+    using Sieve.Exceptions;
+    using Sieve.Models;
+    using Sieve.Services;
+    using System;
+    using System.Linq;
+    using System.Linq.Expressions;
+
+    /// <summary>
+    /// Default implemetation of <see cref="IDynamicFilterableRepository{T}"/>
```
Comment (Reza-Noei) [inline Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs]: * implementation

File: Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    public class DynamicFilteraleRepositoryBase<T> : RepositoryBase<T>, IDynamicFilterableRepository<T> where T : class
```
Comment (Reza-Noei) [inline Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs]: * DynamicFilterableRepositoryBase (update filename also)
tnx

File: Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+// ITNOA
+
+namespace BSN.Commons.Orm.EntityFrameworkCore
+{
+    using BSN.Commons.Extensions;
```
Comment (Reza-Noei) [inline Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs]: why usings are inside the namespace ?

if you did this to prioritize type inference please add some comment for it. 

tnx

File: Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            return query.Paginate(pageNumber, pageSize);
+        }
+
+        /// <inheritdoc />
+        public PagedEntityCollection<T> GetMany(string filters, string sorts, uint pageNumber, uint pageSize)
+        {
+            return GetMany((entity) => true, filters, sorts, pageNumber, pageSize);
+        }
+
+        protected readonly ISieveProcessor sieveProcessor;
```
Comment (Reza-Noei) [inline Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs]: I haven't found any specific convention on `protected` fields in our [`Coding Conventions` page](https://tfs.resaa.net/resaa/Resaa/_wiki/wikis/Resaa.wiki/62/Coding-Conventions).

neither in [Microsoft official guideline ](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/naming-rules)

there is a C# coding style doc written by [google](https://google.github.io/styleguide/csharp-style.html) that says we should write protected fields `_camelcase`

File: Source/BSN.Commons/Infrastructure/IDynamicFilterableRepository.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    /// Add Dynamic filterable query ability to <see cref="IRepository{T}"/> pattern"
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+	public interface IDynamicFilterableRepository<T> : IRepository<T> where T : class
+	{
+        /// <summary>
+        /// Get list of objects based on <paramref name="filters"/>
+        /// </summary>
+        /// <remarks>
+        /// This method retrive all objects based on filters and apply pagination.
```
Comment (Reza-Noei) [inline Source/BSN.Commons/Infrastructure/IDynamicFilterableRepository.cs]: * retrieve 
in line 16 and 17

File: Source/BSN.Commons/Infrastructure/IDynamicFilterableRepository.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        ///      </description>
+        ///  </item>
+        ///  <item>
+        ///      <description>
+        ///         <c>{Operator}</c> is one of the Operators
+        ///      </description>
+        ///  </item>
+        ///  <item>
+        ///      <description>
+        ///          <c>{Value}</c>` is the value to use for filtering
```
Comment (Reza-Noei) [inline Source/BSN.Commons/Infrastructure/IDynamicFilterableRepository.cs]: thank u for expressive comments. 
a character of \` is extra in this line.

Comment (soroshsabz) [reply Source/BSN.Commons/Infrastructure/IDynamicFilterableRepository.cs]: I fixed it.

File: Source/BSN.Commons/Infrastructure/IDynamicFilterableRepository.cs (hunk 5 above)
Comment (soroshsabz) [reply Source/BSN.Commons/Infrastructure/IDynamicFilterableRepository.cs]: I fixed it.

File: Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs (hunk 4 above)
Comment (soroshsabz) [reply Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs]: I fix it.

File: Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs (hunk 3 above)
Comment (soroshsabz) [reply Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs]: I move to outside, for consistency

File: Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs (hunk 2 above)
Comment (soroshsabz) [reply Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs]: Sorry for my mistake

I fixed it.

File: Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs (hunk 1 above)
Comment (soroshsabz) [reply Source/BSN.Commons.Orm.EntityFrameworkCore/DynamicFilteraleRepositoryBase.cs]: I fixed it.

**Review by Reza-Noei — APPROVED** [review]

**Pull request merged**